pkg/provisioner/provisioner.go

```diff
@@ -10,29 +10,34 @@ import (
 Package provisioning defines the API for talking to the provisioning backend.
 */
 
+type ProvisionerFactory interface {
```

Comment:
Consider making this a function-based type rather than an interface, since I think that's all you really need.

```
type ProvisionerFactory func(host *metalkubev1alpha1.BareMetalHost) (Provisioner, error)
```

That'll make usage a little simpler too I think. For example, `ironic.New` can be the factory function, returning an ironic provisioner. It eliminates the need to have a constructor just for your factory.

You can then do things like this:

```
	return &ReconcileBareMetalHost{
		client:             mgr.GetClient(),
		scheme:             mgr.GetScheme(),
		provisionerFactory: ironic.New,
        }
```

Let me know what you think of that approach.

Reply:
I'm working on the ironic version of the factory, and I'm anticipating wanting to have the controller pass in configuration values like the IPA image names, ironic IP address, etc. We could do that with a function-based interface, but it just moves all of those settings to the factory's New() method and means the reconciler has to keep up with them, instead of just keeping up with the factory itself.

Reply:
See https://github.com/dhellmann/baremetal-operator/commit/5f067534702aa3815abeb09250914084c00666f8 for an example of what I mean.

Reply:
If the provisioner factory will ever need multiple methods, then an interface is the way to go. But if it will only ever have the New method, which seems likely, this function-based approach may still pay off. The above example would become:

```
return &ReconcileBareMetalHost{
		client:             mgr.GetClient(),
		scheme:             mgr.GetScheme(),
		provisionerFactory: ironic.NewFactory(x, y, z).New, // renamed to NewFactory just for clarity of the example
        }
```

or you can move the factory into a closure and do this, which is another common pattern in go code:

```
return &ReconcileBareMetalHost{
		client:             mgr.GetClient(),
		scheme:             mgr.GetScheme(),
		provisionerFactory: func(host *metalkubev1alpha1.BareMetalHost) (Provisioner, error) {
				return ironic.New(x, y, z, host)
		},
        }
```

It ends up being quite flexible, and I like that it reduces the "New depth" from `somepackage.New(config).New(host)`

Here is an example of the closure-factory pattern: https://github.com/operator-framework/operator-sdk/blob/v0.5.0/pkg/ansible/runner/runner.go#L193-L196

If you prefer to stick with the interface, as a nit-pick I suggest considering a name like `NewFactory`, since the common assumption is that if I call `thing.New()`, I'll get a new "thing", not a "thing factory".

Reply:
I've renamed the factory creation functions for now. If we want to rewrite this later to be function based, that's fine, but I don't want to spin on that this sprint.